libraries/freertos_plus/standard/freertos_plus_tcp/source/FreeRTOS_DNS.c

```diff
@@ -378,48 +414,63 @@ typedef struct xDNSAnswerRecord DNSAnswerRecord_t;
 	/*-----------------------------------------------------------*/
 
 	/* A DNS reply was received, see if there is any matching entry and
-	call the handler. */
-	static void vDNSDoCallback( TickType_t xIdentifier, const char *pcName, uint32_t ulIPAddress );
-	static void vDNSDoCallback( TickType_t xIdentifier, const char *pcName, uint32_t ulIPAddress )
+	call the handler.  Returns pdTRUE if uxIdentifier was recognised. */
+	static BaseType_t xDNSDoCallback( TickType_t uxIdentifier,
+									  const char *pcName,
+									  uint32_t ulIPAddress );
+	static BaseType_t xDNSDoCallback( TickType_t uxIdentifier,
+									  const char *pcName,
+									  uint32_t ulIPAddress )
 	{
-		const ListItem_t *pxIterator;
-		const MiniListItem_t* xEnd = ( const MiniListItem_t* )listGET_END_MARKER( &xCallbackList );
+	BaseType_t xResult = pdFALSE;
+	const ListItem_t *pxIterator;
+	const MiniListItem_t * xEnd = ( const MiniListItem_t * ) listGET_END_MARKER( &xCallbackList );
 
 		vTaskSuspendAll();
 		{
 			for( pxIterator  = ( const ListItem_t * ) listGET_NEXT( xEnd );
 				 pxIterator != ( const ListItem_t * ) xEnd;
 				 pxIterator  = ( const ListItem_t * ) listGET_NEXT( pxIterator ) )
 			{
-				if( listGET_LIST_ITEM_VALUE( pxIterator ) == xIdentifier )
+				/* The cast will take away the 'configLIST_VOLATILE' */
+				if( uxIdentifier == ( TickType_t ) listGET_LIST_ITEM_VALUE( pxIterator ) )
 				{
-					DNSCallback_t *pxCallback = ( DNSCallback_t * ) listGET_LIST_ITEM_OWNER( pxIterator );
+				DNSCallback_t *pxCallback = ( DNSCallback_t * ) listGET_LIST_ITEM_OWNER( pxIterator );
+
 					pxCallback->pCallbackFunction( pcName, pxCallback->pvSearchID, ulIPAddress );
 					uxListRemove( &pxCallback->xListItem );
 					vPortFree( pxCallback );
+
 					if( listLIST_IS_EMPTY( &xCallbackList ) )
 					{
+						/* The list of outstanding requests is empty. No need for periodic polling. */
 						vIPSetDnsTimerEnableState( pdFALSE );
 					}
+
+					xResult = pdTRUE;
 					break;
 				}
 			}
 		}
 		xTaskResumeAll();
+		return xResult;
 	}
 
-#endif	/* ipconfigDNS_USE_CALLBACKS != 0 */
+#endif /* ipconfigDNS_USE_CALLBACKS == 1 */
 /*-----------------------------------------------------------*/
 
 #if( ipconfigDNS_USE_CALLBACKS == 0 )
-uint32_t FreeRTOS_gethostbyname( const char *pcHostName )
+	uint32_t FreeRTOS_gethostbyname( const char *pcHostName )
```

Comment:
Does this really need to be an either-or scenario?  Why don't we simply have both interface be always available?  The non-callback version could be a simple wrapper for the callback version.

Reply:
At line 308, you will see that the blocking version is also available in case `ipconfigDNS_USE_CALLBACKS == 1` :

    uint32_t FreeRTOS_gethostbyname( const char *pcHostName )
    {
        return FreeRTOS_gethostbyname_a( pcHostName, ( FOnDNSEvent ) NULL, ( void * ) NULL, 0 );
    }
    /*-----------------------------------------------------------*/

Is that what you mean?

Reply:
Yes, thank you.